File: Backend/app/protocols/user_skills.py
from __future__ import annotations

from typing import Any, AsyncIterator, Literal

from pydantic import Field

from app.protocols.ag_ui_action_stream import (
    AgUiActionResult,
    build_ag_ui_action_stream,
)

from app.services.user_skill_documents import (
    CreateUserSkillRequest,
    DeleteUserSkillRequest,
    GetUserSkillRequest,
    SaveUserSkillRequest,
    create_user_skill_document,
    delete_user_skill,
    read_user_skill_document,
    save_user_skill_document,
)
from app.services.user_skill_imports import (
    ImportUserSkillRequest,
    SkillImportError,
    import_user_skill_archive,
)
from app.services.user_skills import (
    ApiModel,
    list_user_skills,
    update_user_skill_enabled,
    user_skill_settings_label,
    user_skills_root_label,
)
from app.services.builtin_skills import (
    builtin_skills_root_label,
    list_builtin_skills,
)


SKILL_CATALOG_EVENT_NAME = "skill-catalog"


class SetUserSkillEnabledRequest(ApiModel):
    """校验用户技能启停动作的 AG-UI 输入。"""

    action: Literal["set-enabled"]
    relative_path: str = Field(min_length=1)
    enabled: bool


def user_skills_capabilities() -> dict[str, Any]:
    """描述独立技能目录 AG-UI 动作和公开状态契约。"""

    return {
        "name": "user-skills",
        "endpoint": "/skills/run",
        "transport": "ag-ui-sse",
        "actions": [
            "list",
            "get",
            "save",
            "create",
            "delete",
            "import",
            "set-enabled",
        ],
        "customEventName": SKILL_CATALOG_EVENT_NAME,
        "stateSnapshotKey": "skillCatalog",
        "root": user_skills_root_label(),
        "builtinRoot": builtin_skills_root_label(),
        "enablement": {
            "default": "enabled",
            "stateFile": user_skill_settings_label(),
            "disabledSkillsField": "disabledSkills",
        },
        "workflowIndependent": True,
    }


def build_user_skills_ag_ui_stream(
    *, payload: dict[str, Any], accept: str | None = None
) -> AsyncIterator[str]:
    """把技能目录动作包装成完整的 AG-UI 生命周期事件流。"""

    skill_input = _skill_catalog_input(payload)
    action = skill_input.get("action")

    async def operation() -> AgUiActionResult:
        if action == "list":
            catalog = list_user_skills()
            builtin_skills = list_builtin_skills()
            result_payload = {
                **catalog.model_dump(by_alias=True, exclude_none=True),
                "builtinRoot": builtin_skills_root_label(),
                "builtinSkills": [
                    skill.model_dump(by_alias=True, exclude_none=True)
                    for skill in builtin_skills
                ],
            }
            message = (
                f"已读取 {len(catalog.skills)} 个用户技能和 "
                f"{len(builtin_skills)} 个内置技能。"
            )
        elif action == "get":
            request = GetUserSkillRequest.model_validate(skill_input)
            document = read_user_skill_document(request.relative_path)
            result_payload = {
                "root": user_skills_root_label(),
                "document": document.model_dump(by_alias=True),
            }
            message = f"已读取技能 {document.name}。"
        elif action == "create":
            request = CreateUserSkillRequest.model_validate(skill_input)
            document = create_user_skill_document(request.content)
            result_payload = {
                "root": user_skills_root_label(),
                "document": document.model_dump(by_alias=True),
            }
            message = f"已创建技能 {document.name}。"
        elif action == "save":
            request = SaveUserSkillRequest.model_validate(skill_input)
            document = save_user_skill_document(
                request.relative_path,
                request.content,
                request.expected_revision,
            )
            result_payload = {
                "root": user_skills_root_label(),
                "document": document.model_dump(by_alias=True),
            }
            message = f"已保存技能 {document.name}。"
        elif action == "delete":
            request = DeleteUserSkillRequest.model_validate(skill_input)
            deleted = delete_user_skill(request.relative_path)
            result_payload = {
                "root": user_skills_root_label(),
                "deleted": deleted.model_dump(by_alias=True),
            }
            message = f"已删除技能 {deleted.name}。"
        elif action == "import":
            request = ImportUserSkillRequest.model_validate(skill_input)
            imported = import_user_skill_archive(
                request.file_name,
                request.archive_base64,
            )
            result_payload = imported.model_dump(by_alias=True, exclude_none=True)
            message = f"已导入技能 {imported.imported.name}。"
        elif action == "set-enabled":
            request = SetUserSkillEnabledRequest.model_validate(skill_input)
            skill = update_user_skill_enabled(
                request.relative_path,
                request.enabled,
            )
            result_payload = {
                "root": user_skills_root_label(),
                "skill": skill.model_dump(by_alias=True, exclude_none=True),
            }
            status_label = "开启" if request.enabled else "关闭"
            message = f"已{status_label}技能 {skill.name}。"
        else:
            raise ValueError(
                "skillCatalog.action 必须是 list、get、save、create、delete、import "
                "或 set-enabled。"
            )
        return AgUiActionResult(
            data={"action": action, **result_payload},
            message=message,
        )

    return build_ag_ui_action_stream(
        payload=payload,
        event_name=SKILL_CATALOG_EVENT_NAME,
        state_key="skillCatalog",
        run_id_prefix="skills",
        operation=operation,
        error_message_prefix="技能操作失败",
        error_data=lambda exc: {
            "action": action,
            **(
                {"code": exc.code}
                if isinstance(exc, SkillImportError)
                else {}
            ),
        },
        accept=accept,
    )


def _skill_catalog_input(payload: dict[str, Any]) -> dict[str, Any]:
    """从 AG-UI forwardedProps 中提取技能目录动作输入。"""

    forwarded_props = payload.get("forwardedProps")
    if not isinstance(forwarded_props, dict):
        return {}
    skill_catalog = forwarded_props.get("skillCatalog")
    return skill_catalog if isinstance(skill_catalog, dict) else {}

File: prototype/src/renderer/src/service/userSkills.ts
import type {
  BuiltinSkill,
  UserSkill,
  UserSkillCatalog,
  UserSkillDocument,
  UserSkillIssue
} from '../typings'
import { runAgUiAction, type AgUiPayloadEnvelope } from './agUiClient'

type SkillCatalogAgUiPayload = AgUiPayloadEnvelope & {
  status: 'completed' | 'failed'
  action?: 'list' | 'get' | 'save' | 'create' | 'delete' | 'import' | 'set-enabled'
  root?: string
  skills?: UserSkill[]
  builtinRoot?: string
  builtinSkills?: BuiltinSkill[]
  skippedCount?: number
  issues?: UserSkillIssue[]
  document?: UserSkillDocument
  deleted?: { name: string; relativePath: string }
  imported?: UserSkill
  skill?: UserSkill
  error?: { type?: string; message?: string }
}

const SKILL_STATUS_LIST = ['completed', 'failed'] as const
const SKILL_CATALOG_EVENT = 'skill-catalog'
const SKILL_CATALOG_KEY = 'skillCatalog'

function getSkillCatalogUrl(): string {
  /** 根据桌面运行时配置解析技能 AG-UI 地址。 */
  const agentBaseUrl = window.xcodeAgent?.agentBaseUrl
  return agentBaseUrl ? `${agentBaseUrl.replace(/\/$/, '')}/skills/run` : '/api/agent/skills/run'
}

async function runSkillCatalogAgent(
  input: Record<string, unknown>,
  messageContent: string
): Promise<SkillCatalogAgUiPayload> {
  /** 运行一次独立技能动作，payload 合并/校验/失败抛错统一由 runAgUiAction 处理。 */
  return runAgUiAction<SkillCatalogAgUiPayload>({
    url: getSkillCatalogUrl(),
    message: messageContent,
    eventName: SKILL_CATALOG_EVENT,
    stateKey: SKILL_CATALOG_KEY,
    forwardedProps: { [SKILL_CATALOG_KEY]: input },
    statusList: SKILL_STATUS_LIST,
    emptyMessage: '技能接口没有返回有效的 AG-UI 状态。',
    failedMessage: '技能操作失败。'
  })
}

export async function requestUserSkills(): Promise<UserSkillCatalog> {
  /** 同时读取用户技能、内置技能和用户启用状态。 */
  const catalog = await runSkillCatalogAgent({ action: 'list' }, '读取本地用户技能列表。')
  if (
    typeof catalog.root !== 'string' ||
    !Array.isArray(catalog.skills) ||
    typeof catalog.builtinRoot !== 'string' ||
    !Array.isArray(catalog.builtinSkills) ||
    !Array.isArray(catalog.issues)
  ) {
    throw new Error('技能接口返回的数据结构无效。')
  }
  return {
    root: catalog.root,
    skills: catalog.skills,
    builtinRoot: catalog.builtinRoot,
    builtinSkills: catalog.builtinSkills,
    skippedCount: Number(catalog.skippedCount || 0),
    issues: catalog.issues
  }
}

export async function requestUserSkillDocument(relativePath: string): Promise<UserSkillDocument> {
  /** 读取一个可编辑用户技能的完整 SKILL.md。 */
  const response = await runSkillCatalogAgent(
    { action: 'get', relativePath },
    `读取用户技能 ${relativePath}。`
  )
  if (!response.document) throw new Error('技能接口没有返回完整内容。')
  return response.document
}

export async function createUserSkillDocument(input: {
  content: string
}): Promise<UserSkillDocument> {
  /** 创建默认开启的用户技能文档。 */
  const response = await runSkillCatalogAgent({ action: 'create', ...input }, '创建用户技能。')
  if (!response.document) throw new Error('技能接口没有返回创建结果。')
  return response.document
}

export async function saveUserSkillDocument(input: {
  relativePath: string
  content: string
  expectedRevision: string
}): Promise<UserSkillDocument> {
  /** 按内容 revision 乐观保存用户技能。 */
  const response = await runSkillCatalogAgent(
    { action: 'save', ...input },
    `保存用户技能 ${input.relativePath}。`
  )
  if (!response.document) throw new Error('技能接口没有返回保存结果。')
  return response.document
}

export async function deleteUserSkill(relativePath: string): Promise<void> {
  /** 删除用户技能及其后端启用状态。 */
  const response = await runSkillCatalogAgent(
    { action: 'delete', relativePath },
    `删除用户技能 ${relativePath}。`
  )
  if (response.deleted?.relativePath !== relativePath) {
    throw new Error('技能接口没有返回有效的删除结果。')
  }
}

export async function importUserSkillArchive(input: {
  archiveBase64: string
  fileName: string
}): Promise<UserSkill> {
  /** 通过 AG-UI 导入默认开启的用户技能 ZIP。 */
  const response = await runSkillCatalogAgent(
    { action: 'import', ...input },
    `导入用户技能压缩包 ${input.fileName}。`
  )
  if (!response.imported) throw new Error('技能接口没有返回有效的导入结果。')
  return response.imported
}

export async function setUserSkillEnabled(input: {
  relativePath: string
  enabled: boolean
}): Promise<UserSkill> {
  /** 持久化用户技能启用状态并返回服务端确认后的摘要。 */
  const response = await runSkillCatalogAgent(
    { action: 'set-enabled', ...input },
    `${input.enabled ? '开启' : '关闭'}用户技能 ${input.relativePath}。`
  )
  if (!response.skill) throw new Error('技能接口没有返回有效的启用状态。')
  return response.skill
}
